broker/lib/fabrik/DirectorManager.js

```diff
@@ -177,7 +189,179 @@ class DirectorManager extends BaseManager {
+
+  removeCachedTask(serviceInstanceId) {
+    return boshOperationCache.deleteBoshTask(serviceInstanceId);
+  }
+
+  getCurrentOperationState(serviceInstanceId) {
+    let output = {
+      'cached': false,
+      'task_id': null
+    };
+
+    return Promise.all([boshOperationCache.containsServiceInstance(serviceInstanceId), boshOperationCache.getBoshTask(serviceInstanceId)])
+      .spread((cached, taskId) => {
+        output.cached = cached;
+        output.task_id = taskId;
+      })
+      .return(output);
+  }
+
   createOrUpdateDeployment(deploymentName, params, args) {
+    logger.info(`Checking rate limits against bosh for deployment `);
+    const previousValues = _.get(params, 'previous_values');
+    const action = _.isPlainObject(previousValues) ? CONST.OPERATION_TYPE.UPDATE : CONST.OPERATION_TYPE.CREATE;
+    const scheduled = _.get(params, 'scheduled') || false;
+    const dbUpdate = _.get(params, '_serviceFabrikDbUpdate') || false;
+    const policyApplicable = config.enable_bosh_rate_limit;
+    _.omit(params, 'scheduled');
+    _.omit(params, '_serviceFabrikDbUpdate');
+
+    const decisionMaker = {
+      'shouldRunNow': false,
+      'cached': false,
+      'dbUpdate': false,
+      'policyApplicable': config.enable_bosh_rate_limit,
+      'directorError': false
+    };
+
+    return this.executePolicy(scheduled, action, deploymentName, dbUpdate, policyApplicable)
+      .then(checkResults => {
+        if (checkResults.directorError) {
+          decisionMaker.directorError = true;
+          return Promise.resolve(true);
+        }
+        const shouldRunNow = checkResults.shouldRunNow;
+
+        if (!decisionMaker.policyApplicable) {
+          return Promise.resolve();
+        }
+        if (dbUpdate) {
+          decisionMaker.shouldRunNow = true;
+          decisionMaker.dbUpdate = true;
+          return Promise.resolve();
+        }
+        if (shouldRunNow) {
+          decisionMaker.shouldRunNow = shouldRunNow;
+          return boshOperationCache.containsDeployment(deploymentName).then(inCache => {
+            if (inCache) {
+              decisionMaker.cached = true;
+            }
+            return Promise.resolve();
+          });
+        }
+        if (!scheduled) {
+          return Promise.resolve(true);
+        }
+        throw new DeploymentDelayed(deploymentName);
+      })
+      .then((storeInCache) => {
+        if (storeInCache) {
+          // stagger here by putting it into etcd cache and return promise
+          return boshOperationCache.store(this.plan.id, deploymentName, params, args)
+            .then(() => {
+              decisionMaker.cached = true;
+              return Promise.resolve();
+            })
+            .catch(err => {
+              logger.error('Error updating etcd cache', err);
+              throw new CacheUpdateError(deploymentName);
+            });
+        }
+      })
+      .then(() => {
+        if (decisionMaker.shouldRunNow && decisionMaker.cached) {
```

Comment:
But once queued operation is complete, its anyways cleared. when will this occur? is this to handle failure in delete of queued entries? if such was the case, then it would be better to have a timestamp when the entry was queued and ensure its overrwritten or invalidated instead of explicit delete like the below to handle delete failure.

Reply:
No, this is for the case where the task poller runs, finds that slots are available in BOSH - then, the entry from etcd for the deployment should be removed and then processed. Note that this happens _before_ the task is submitted to BOSH

Reply:
ok